perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_user_attribute_def_virt_sponsoredMembershipInOrganizations.java

```diff
@@ -0,0 +1,100 @@
+package cz.metacentrum.perun.core.impl.modules.attributes;
+
+import cz.metacentrum.perun.core.api.Attribute;
+import cz.metacentrum.perun.core.api.AttributeDefinition;
+import cz.metacentrum.perun.core.api.AttributesManager;
+import cz.metacentrum.perun.core.api.Group;
+import cz.metacentrum.perun.core.api.Member;
+import cz.metacentrum.perun.core.api.Status;
+import cz.metacentrum.perun.core.api.User;
+import cz.metacentrum.perun.core.api.exceptions.AttributeNotExistsException;
+import cz.metacentrum.perun.core.api.exceptions.InternalErrorException;
+import cz.metacentrum.perun.core.api.exceptions.WrongAttributeAssignmentException;
+import cz.metacentrum.perun.core.bl.GroupsManagerBl;
+import cz.metacentrum.perun.core.impl.PerunSessionImpl;
+import cz.metacentrum.perun.core.implApi.modules.attributes.SkipValueCheckDuringDependencyCheck;
+import cz.metacentrum.perun.core.implApi.modules.attributes.UserVirtualAttributesModuleAbstract;
+import cz.metacentrum.perun.core.implApi.modules.attributes.UserVirtualAttributesModuleImplApi;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Get all records (as list) of attribute group:def:sponsorOrganizationIdentifier
+ * from any Group where user is valid Member.
+ *
+ * @author Michal Stava stavamichal@gmail.com
+ */
+@SkipValueCheckDuringDependencyCheck
+public class urn_perun_user_attribute_def_virt_sponsoredMembershipInOrganizations extends UserVirtualAttributesModuleAbstract implements UserVirtualAttributesModuleImplApi {
+
+	private static final String sponsorOrganizationIdentifierFriendlyName = "sponsorOrganizationIdentifier";
+
+	private static final String A_G_D_sponsorOrganizationIdentifier = AttributesManager.NS_GROUP_ATTR_DEF + ":" + sponsorOrganizationIdentifierFriendlyName;
+
+	private final static Logger log = LoggerFactory.getLogger(urn_perun_user_attribute_def_virt_sponsoredMembershipInOrganizations.class);
+
+	@Override
+	public Attribute getAttributeValue(PerunSessionImpl sess, User user, AttributeDefinition attributeDefinition) {
+
+		Attribute attribute = new Attribute(attributeDefinition);
+
+		attribute.setValue(getSponsorOrganizationIdentifiersFromGroups(sess, user));
+
+		return attribute;
+	}
+
+	/**
+	 * Collect sponsor organization's identifiers from perun Groups
+	 * @param sess Perun session
+	 * @param user User for whom the values should be collected
+	 * @return List of collected identifiers
+	 * @throws InternalErrorException When some error occurs, see exception cause for details.
+	 */
+	private List<String> getSponsorOrganizationIdentifiersFromGroups(PerunSessionImpl sess, User user) {
```

Comment:
NIT: This is just to show, how this could be done with the Java8 stream API (even though I know you will probably not like it :wink:) If you would like to, you can implement it like this.
```
	private List<String> getSponsorOrganizationIdentifiersFromGroups(PerunSessionImpl sess, User user) {
		return sess.getPerunBl().getMembersManagerBl().getMembersByUserWithStatus(sess, user, Status.VALID).stream()
			.flatMap(m -> sess.getPerunBl().getGroupsManagerBl().getGroupsWhereMemberIsActive(sess, m).stream())
			.map(g -> getOrganizationIdentifierAttribute(sess, g, user))
			.filter(Objects::nonNull)
			.map(Attribute::valueAsString)
			.filter(Objects::nonNull)
			.distinct()
			.collect(Collectors.toList());
	}

	private Attribute getOrganizationIdentifierAttribute(PerunSessionImpl sess, Group group, User user) {
		try {
			return sess.getPerunBl().getAttributesManagerBl().getAttribute(sess, group, A_G_D_sponsorOrganizationIdentifier);
		} catch (WrongAttributeAssignmentException e) {
			throw new InternalErrorException("Wrong assignment of " + A_G_D_sponsorOrganizationIdentifier + " for user " + user.getId(), e);
		} catch (AttributeNotExistsException e) {
			log.debug("Attribute " + A_G_D_sponsorOrganizationIdentifier + " of group " + group.getId() + " does not exist, values will be skipped", e);
		}
		return null;
	}
```

Reply:
Why not. Lets go to the future :sweat_smile: Changed.